src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -55,55 +52,76 @@ impl MinOverlappingPicker {
         target_tables: &[SstableInfo],
         level_handlers: &[LevelHandler],
     ) -> (Vec<SstableInfo>, Vec<SstableInfo>) {
-        let mut scores = vec![];
-        for left in 0..select_tables.len() {
-            if level_handlers[self.level].is_pending_compact(&select_tables[left].sst_id) {
+        let mut select_file_ranges = vec![];
+        for (idx, sst) in select_tables.iter().enumerate() {
+            if level_handlers[self.level].is_pending_compact(&sst.sst_id) {
                 continue;
             }
             let mut overlap_info = self.overlap_strategy.create_overlap_info();
-            let mut select_file_size = 0;
-            for (right, table) in select_tables.iter().enumerate().skip(left) {
-                if level_handlers[self.level].is_pending_compact(&table.sst_id) {
+            overlap_info.update(sst);
+            let overlap_files_range = overlap_info.check_multiple_overlap(target_tables);
+
+            if overlap_files_range.is_empty() {
+                return (vec![sst.clone()], vec![]);
+            }
+            select_file_ranges.push((idx, overlap_files_range));
+        }
+        select_file_ranges.retain(|(_, range)| {
+            let mut pending_compact = false;
+            for other in &target_tables[range.clone()] {
+                if level_handlers[self.target_level].is_pending_compact(&other.sst_id) {
+                    pending_compact = true;
                     break;
                 }
-                if select_file_size > self.max_select_bytes {
+            }
+            !pending_compact
+        });
+
+        let mut min_score = u64::MAX;
+        let mut min_score_select_range = 0..0;
+        let mut min_score_target_range = 0..0;
+        let mut min_score_select_file_size = 0;
+        for left in 0..select_file_ranges.len() {
+            let mut select_file_size = 0;
+            let mut target_level_overlap_range = select_file_ranges[left].1.clone();
+            let mut total_file_size = 0;
+            for other in &target_tables[target_level_overlap_range.clone()] {
+                total_file_size += other.file_size;
+            }
+            let left_idx = select_file_ranges[left].0;
+            for (idx, range) in select_file_ranges.iter().skip(left) {
+                if select_file_size > self.max_select_bytes
+                    || range.start >= target_level_overlap_range.end
```

Comment:
Is it possible the fist select_file_ranges is left eq right, which makes `range.start >= target_level_overlap_range.end` true and thus the first file can never be selected?

Reply:
No. Because the range of target_level_overlap_range could never be empty (left eq right). We have return empty range as a trivial-move task before.